### PR TITLE
Flow-sensitive typing for `equal?` Fixes #3456

### LINKED ISSUE #3456: Flow-sensitive typing for `Object#equal?`
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20Module%0A%20%20include%20T%3A%3ASig%0Aend%0A%0Aclass%20Main%0A%20%20sig%20%7Bparams(x%3A%20T.any(Integer%2C%20String)).void%7D%0A%20%20def%20self.test1(x)%0A%20%20%20%20if%20x.equal%3F(0)%0A%20%20%20%20%20%20T.reveal_type(x)%20%23%20not%20working%20(Integer)%0A%20%20%20%20else%0A%20%20%20%20%20%20T.reveal_type(x)%20%23%20working%20ok%20(T.any(Integer%2C%20String))%0A%20%20%20%20end%0A%20%20end%0Aend)

```ruby
# typed: true
class Module
  include T::Sig
end

class Main
  sig {params(x: T.any(Integer, String)).void}
  def self.test1(x)
    if x.equal?(0)
      T.reveal_type(x) # not working (Integer)
    else
      T.reveal_type(x) # working ok (T.any(Integer, String))
    end
  end
end
```

#### Observed output

```
editor.rb:12: Revealed type: T.any(Integer, String) https://srb.help/7014
    12 |      T.reveal_type(x) # working ok (T.any(Integer, String))
              ^^^^^^^^^^^^^^^^
  From:
    editor.rb:8:
     8 |  def self.test1(x)
                         ^

editor.rb:10: Revealed type: T.any(Integer, String) https://srb.help/7014
    10 |      T.reveal_type(x) # not working (Integer)
              ^^^^^^^^^^^^^^^^
  From:
    editor.rb:8:
     8 |  def self.test1(x)
                         ^
Errors: 2
```

#### Expected behavior

If a value is reference equal to another value with a known type, we should be able to narrow the type of that value to the known type.

This is already implemented when the method `Object#==` is called, but not when `Object#equal?` is called.

`==` and `equal?` are the same method if not overridden. Ruby recommends overriding `==`  if the class wants to provide a custom notion of equality, while `equal?` should never be overridden so there is always still a way to check for reference equality.

(An unfortunate limitation of the way this is handled in the code right now is that Sorbet doesn't actually check whether the method named `==` or `equal?` corresponds to the one defined on Object or something implemented by a derived class, but that's not something we need to fix to call this bug fixed.)

- - -

Note that if the type test fails, we can't necessarily update the type. `!x.equal?(0)` does not imply that `x` is not an Integer because there are many other valid values that `x` could be and still have type `Integer`.

### PR DESCRIPTION
**Pull request by sharpobject**
Just treat `equal?` the same way as `==`. Per the very helpful description over at #3456, using `equal?` in this way is actually more reliable than using `==`, because by convention people override `==` and not `equal?`.

### Motivation
#3456

### Test plan
See included automated tests.

### CODE REVIEW

File: infer/environment.cc [hunk 1]
```diff
@@ -437,16 +437,18 @@ void Environment::updateKnowledge(core::Context ctx, cfg::LocalRef local, core::
...
         if (!knowledgeFilter.isNeeded(local)) {
             return;
         }
         auto &whoKnows = getKnowledge(local);
         const auto &argType = send->args[0].type;
         const auto &recvType = send->recv.type;
 
-        auto &truthy = send->fun == core::Names::eqeq() ? whoKnows.truthy : whoKnows.falsy;
-        auto &falsy = send->fun == core::Names::eqeq() ? whoKnows.falsy : whoKnows.truthy;
+        bool funIsEq = send->fun == core::Names::eqeq() || send->fun == core::Names::equal_p();
```
Comment (sharpobject) [inline infer/environment.cc:449]: You could also do something like
```cpp
auto &truthy = send->fun == core::Names::neq() ? whoKnows.falsy : whoKnows.truthy;
```
but I thought the extra line was preferable to the double negative.

File: test/testdata/infer/control_flow/bug_3456.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# typed: true
```
Comment (jez) [inline test/testdata/infer/control_flow/bug_3456.rb:1]: Would you mind naming this test according to what it tests not what bug number it fixes? like `control_flow/equal_p.rb` or something?

**Review by jez — APPROVED** [review]
Looks good, thanks!

File: test/testdata/infer/control_flow/bug_3456.rb (hunk 2 above)
Comment (sharpobject) [reply test/testdata/infer/control_flow/bug_3456.rb]: Alright.

**Comment by sharpobject** [comment]
Do you need anything else for this one?

**Pull request merged**